allenact/embodiedai/sensors/vision_sensors.py

```diff
@@ -0,0 +1,264 @@
+                high=np.float32(unnormalized_supremum),
+                shape=shape,
+            )
+        else:
+            out_shape = shape[:-1] + (1,)
+            low = np.tile(
+                (unnormalized_infimum - cast(np.ndarray, self._norm_means))
+                / cast(np.ndarray, self._norm_sds),
+                out_shape,
+            )
+            high = np.tile(
+                (unnormalized_supremum - cast(np.ndarray, self._norm_means))
+                / cast(np.ndarray, self._norm_sds),
+                out_shape,
+            )
+            return gym.spaces.Box(low=np.float32(low), high=np.float32(high))
+
+    def _get_observation_space(self):
+        return self._observation_space
+
+    @property
+    def height(self) -> Optional[int]:
+        """Height that input image will be rescale to have.
+
+        # Returns
+
+        The height as a non-negative integer or `None` if no rescaling is done.
+        """
+        return self._height
+
+    @property
+    def width(self) -> Optional[int]:
+        """Width that input image will be rescale to have.
+
+        # Returns
+
+        The width as a non-negative integer or `None` if no rescaling is done.
+        """
+        return self._width
+
+    @abstractmethod
+    def frame_from_env(self, env: EnvType, task: Optional[SubTaskType]) -> np.ndarray:
+        raise NotImplementedError
+
+    def get_observation(
+        self, env: EnvType, task: Optional[SubTaskType], *args: Any, **kwargs: Any
+    ) -> Any:
+        im = self.frame_from_env(env=env, task=task)
+        assert (
+            im.dtype == np.float32 and (len(im.shape) == 2 or im.shape[-1] == 1)
+        ) or (im.shape[-1] == 3 and im.dtype == np.uint8), (
+            "Input frame must either have 3 channels and be of"
+            " type np.uint8 or have one channel and be of type np.float32"
+        )
+
+        if self._scale_first:
+            if self.scaler is not None and im.shape[:2] != (self._height, self._width):
+                im = np.array(self.scaler(self.to_pil(im)), dtype=im.dtype)  # hwc
+
+        assert im.dtype in [np.uint8, np.float32]
+
+        if im.dtype == np.uint8:
+            im = im.astype(np.float32) / 255.0
+
+        if self._should_normalize:
+            im -= self._norm_means
+            im /= self._norm_sds
+
+        if not self._scale_first:
+            if self.scaler is not None and im.shape[:2] != (self._height, self._width):
+                im = np.array(self.scaler(self.to_pil(im)), dtype=np.float32)  # hwc
+
+        return im
+
+
+class RGBSensor(VisionSensor[EnvType, SubTaskType], ABC):
+    def __init__(
+        self,
+        use_resnet_normalization: bool = False,
+        mean: Optional[np.ndarray] = np.array(
+            [[[0.485, 0.456, 0.406]]], dtype=np.float32
+        ),
+        stdev: Optional[np.ndarray] = np.array(
+            [[[0.229, 0.224, 0.225]]], dtype=np.float32
+        ),
```

Comment:
I realize now that this isn't the best style as it leads to some weird results:
```
import allenact.embodiedai.sensors.vision_sensors as vs
class S(vs.RGBSensor):
    def frame_from_env(*args, **kwargs):
        pass
s0 = S(height=10, width=10, use_resnet_normalization=True)
s0._norm_means[0,0,0] = 3

s1 = S(height=10, width=10, use_resnet_normalization=True)
print(s1._norm_means) # Prints "[[[3.    0.456 0.406]]]" which seems bad
```
Can you add a TODO to fix this?

Reply:
Good catch!